lib/src/bloc_comparative_builder.dart

```diff
@@ -0,0 +1,44 @@
+import 'package:bloc/bloc.dart';
+import 'package:flutter_hooks/flutter_hooks.dart';
+import 'package:hooked_bloc/hooked_bloc.dart';
+import 'package:hooked_bloc/src/config/hooked_bloc_config.dart';
+
+typedef BlocComparativeBuilderCondition<S> = bool Function(
+  S previous,
+  S current,
+);
+
+/// Returns current state of BlocBase class basing on state's comparison
+///
+/// The [useBlocComparativeBuilder] will filter any state,
+/// via specified [buildWhen] parameter.
+///
+/// For building bloc's state without condition, see [useBlocBuilder].
+///
+/// [buildWhen] local filter function, that will filter incoming states
+/// from [BlocBase.stream] with access to the previous and current value.
+S useBlocComparativeBuilder<C extends BlocBase, S>(
+  BlocBase<S> bloc, {
+  required BlocComparativeBuilderCondition<S> buildWhen,
+}) {
+  // The stream doesn't support filtering with the previous and current value
+  // We have to manually store previous value, especially for initial state
+  final currentState = useState(bloc.state);
+
+  final state = useMemoized(
+    () => bloc.stream.where(
+      (nextState) {
+        final shouldRebuild = buildWhen(currentState.value, nextState);
+        currentState.value = nextState;
+        return shouldRebuild;
+      },
+    ),
+    [bloc],
+  );
+
+  return useStream(
+    state,
+    initialData: bloc.state,
+    preserveState: false,
+  ).requireData!;
+}
```

Comment:
It will build widget two times every new state. First on stream event, and second when changing state
```suggestion
  final currentState = useState(bloc.state);

  final state = useMemoized(
    () => bloc.stream.comparativeWhere(buildWhen, initialValue: bloc.value),
    [bloc],
  );

  return useStream(
    state,
    initialData: bloc.state,
    preserveState: false,
  ).requireData!;
}

class ComparativeWhereStreamTransformer<S> extends StreamTransformerBase<S, S> {
  const ComparativeWhereStreamTransformer(this.comparator, {this.initialValue});

  final bool Function(S, S) comparator;
  final S? initialValue;

  @override
  Stream<S> bind(Stream<S> stream) async* {
    S? previousValue = initialValue;

    await for (final value in stream) {
      if (previousValue != null && comparator(previousValue, value)) {
        yield value;
      }
      previousValue = value;
    }
  }
}

extension StreamExt<T> on Stream<T> {
  Stream<T> comparativeWhere(
    bool Function(T, T) comparator, {
    T? initialValue,
  }) =>
      ComparativeWhereStreamTransformer(comparator, initialValue: initialValue)
          .bind(this);
}

```

Reply:
I have corrected the code with `useRef`. Thanks.